src/ring_theory/roots_of_unity.lean

```diff
@@ -581,6 +621,88 @@ begin
+  have hkl : k = l,
+  { apply nat.dvd_antisymm,
+    { have hpowk := ((is_primitive_root.iff_def z l).1 ((mem_primitive_roots hl).1 hz.2)).1,
+      exact ((is_primitive_root.iff_def z k).1 ((mem_primitive_roots hk).1 hz.1)).2 l hpowk },
+    { have hpowl := ((is_primitive_root.iff_def z k).1 ((mem_primitive_roots hk).1 hz.1)).1,
+      exact ((is_primitive_root.iff_def z l).1 ((mem_primitive_roots hl).1 hz.2)).2 k hpowl } },
+  exfalso,
+  exact h hkl
+end
+
+/-- If there is a `n`-th primitive root of unity in `R` and `b` divides `n`,
+then there is a `b`-th primitive root of unity in `R`. -/
+lemma pow {ζ : R} {n : ℕ} {a b : ℕ} 
+  (hn: 0 < n) (h : is_primitive_root ζ n) (hprod : n = a * b) :
+  is_primitive_root (ζ ^ a) b :=
+begin
+  apply (iff_def (ζ ^ a) b).2,
+  split,
+  { calc (ζ ^ a) ^ b
+       = ζ ^ n : by rw [← pow_mul', mul_comm, hprod]
+   ... = 1 : by rw ((iff_def ζ n).1 h).1 },
+  { intros l hl,
+    rw [← pow_mul', mul_comm] at hl,
+    have hdiv := ((iff_def ζ n).1 h).2 (a * l) hl,
+    rw hprod at hdiv,
+    have hzeroa : a ≠ 0,
+    { by_contra hz,
+      simp only [not_not] at hz,
+      simp only [hz, zero_mul] at hprod,
+      exact ne.symm (ne_of_lt hn) hprod },
+    exact (mul_dvd_mul_iff_left hzeroa).1 hdiv }
+end
+
+/-- `nth_roots n` as a `finset` is equal to the union of `primitive_roots i R` for `i ∣ n`
+if there is a primitive root of unity in `R`. -/
+lemma nth_roots_one_eq_bind_primitive_roots {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) :
+  nth_roots_finset n R = finset.bind (nat.divisors ↑n) (λ (i : ℕ), (primitive_roots i R)) :=
+begin
+  rw [nth_roots_finset, ← multiset.to_finset_eq (nth_roots_nodup h)],
+  symmetry,
+  apply finset.eq_of_subset_of_card_le,
+  { intros x hx,
+    simp only [exists_prop, finset.mem_bind, finset.mem_filter, finset.mem_range] at hx,
+    cases hx with a ha,
+    simp only [mem_nth_roots, pnat.pos, finset.mem_mk],
+    simp only [nat.mem_divisors, and_true, ne.def, pnat.ne_zero, not_false_iff] at ha,
+    obtain ⟨d, hd⟩ := ha.1,
+    have hazero : 0 < a,
+    { by_contra h,
+      simp only [le_zero_iff_eq, not_lt] at h,
+      simp only [h, zero_mul, pnat.ne_zero] at hd,
+      exact hd },
+    calc  x ^ ↑n
+        = x ^ (a * d)  : by rw hd
+    ... = (x ^ a) ^ d  : pow_mul x a d
+    ... = 1 ^ d        : by rw ((mem_primitive_roots hazero).1 ha.2).1
+    ... = 1            : one_pow d },
+  { simp only [card_mk],
+    rw card_nth_roots h,
+    apply le_of_eq,
+    rw finset.card_bind,
+    { rw @finset.sum_congr _ _ (nat.divisors ↑n) (finset.filter (λ (x : ℕ), x ∣ n)
+    (finset.range (n : ℕ).succ))  (λ i, (primitive_roots i R).card)
+    (λ i, nat.totient i) _ (nat.finset_div_eq_filter n) _,
+    { symmetry; exact nat.sum_totient n },
+    { intros x hx,
+      simp only [finset.mem_filter, finset.mem_range] at hx,
+      simp,
+      obtain ⟨d, hd⟩ := hx.2,
+      rw mul_comm at hd,
+      exact card_primitive_roots (pow (pnat.pos n) h hd) (pnat.pos_of_div_pos hx.2) } },
```

Comment:
if I am mentally parsing correctly, should lines 692-700 inclusive have an extra indent?